### PR TITLE
Implement WebGL GetRenderbufferParameter

### LINKED ISSUE #20514: Implement WebGLRenderingContext.getRenderbufferParameter
https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.7

### PR DESCRIPTION
**Pull request by fnune**
This needed a bump of gleam to version 0.4.33 for this https://github.com/servo/gleam/pull/162

I think my changes in test expectations are pretty naive and I have to wait for the PR to run on CI to see what the actual impact is. I'd like some guidance on this, too.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach build-geckolib` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #20514 (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20631)

### CODE REVIEW

**Comment by brainlessdeveloper** [comment]
https://github.com/servo/gleam/pull/162

**Comment by brainlessdeveloper** [comment]
r?

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -2345,6 +2345,41 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+        if !target_matches || !pname_matches {
+            self.webgl_error(InvalidEnum);
+            return NullValue();
+        }
+
+        let (sender, receiver) = webgl_channel().unwrap();
+        self.send_command(WebGLCommand::GetRenderbufferParameter(target, pname, sender));
+
+        match receiver.recv().unwrap() {
+            value if value != 0 => Int32Value(value),
```
Comment (gootorov) [inline components/script/dom/webglrenderingcontext.rs]: `0` can be a valid result, right?. I think it should be just `Int32Value(receiver.recv().unwrap())`.

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -2345,6 +2345,41 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
         Int32Value(receiver.recv().unwrap())
     }
 
+    #[allow(unsafe_code)]
+    // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.7
+    unsafe fn GetRenderbufferParameter(&self, _cx: *mut JSContext, target: u32, pname: u32) -> JSVal {
```
Comment (gootorov) [inline components/script/dom/webglrenderingcontext.rs]: For functions with many arguments, we usually indent like so
```Rust
unsafe fn GetRenderbufferParameter(
    &self,
    ...
) -> JSVal {
```

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -2345,6 +2345,41 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+        };
+
+        if !target_matches || !pname_matches {
+            self.webgl_error(InvalidEnum);
+            return NullValue();
+        }
+
+        let (sender, receiver) = webgl_channel().unwrap();
+        self.send_command(WebGLCommand::GetRenderbufferParameter(target, pname, sender));
+
```
Comment (gootorov) [inline components/script/dom/webglrenderingcontext.rs]: The following check is missing:
`If  the  renderbuffer  currently  bound  to
target
is  zero,   then
INVALID_-
OPERATION
is generated.`

Comment (gootorov) [reply components/script/dom/webglrenderingcontext.rs]: It's just this, basically :)
```Rust
if self.bound_renderbuffer.get().is_none() {
    self.webgl_error(InvalidOperation);
    return NullValue();
}
```

**Comment by brainlessdeveloper** [comment]
I'll get to this in the evening, thanks :) Also I think my expectations for test failures are wrong, I should be updating more `.ini` files.

**Comment by brainlessdeveloper** [comment]
I think I just need to update test failure expectations now?

**Comment by nox** [comment]
Many more tests now fail in `/_mozilla/webgl/conformance-2.0.0/conformance2/context/methods-2.html` but that's just because the test is running further AFAICT. Update those expectations and we will be able to merge this PR.

File: tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/context/methods.html.ini [hunk 4]
```diff
@@ -1,7 +1,4 @@
 [methods.html]
-  [WebGL test #0: Property either does not exist or is not a function: getRenderbufferParameter]
-    expected: FAIL
-
   [WebGL test #1: Property either does not exist or is not a function: getUniform]
```
Comment (gootorov) [inline tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/context/methods.html.ini]: This index should be updated too.

File: components/canvas/webgl_thread.rs [hunk 5]
```diff
@@ -1188,6 +1190,17 @@ impl WebGLImpl {
...
 
+    // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.7
+    fn get_renderbuffer_parameter(
+        gl: &gl::Gl,
+        target: u32,
+        pname: u32,
+        chan: WebGLSender<i32>
+    ) {
+        let parameter = gl.get_renderbuffer_parameter_iv(target, pname);
+        chan.send(parameter).unwrap()
```
Comment (gootorov) [inline components/canvas/webgl_thread.rs]: nit: Missing semicolon.

**Comment by brainlessdeveloper** [comment]
Right, fixed. r?

**Comment by brainlessdeveloper** [comment]
There are more tests failures that my changes don't address:
```
  FAIL WebGL test #57: gl.getUniform(boolProgram, bvalLoc) should be true. Threw exception TypeError: gl.getUniform is not a function                                                                            
  FAIL WebGL test #58: gl.getUniform(boolProgram, bval2Loc) should be true,false. Threw exception TypeError: gl.getUniform is not a function                                                                     
  FAIL WebGL test #59: gl.getUniform(boolProgram, bval3Loc) should be true,false,true. Threw exception TypeError: gl.getUniform is not a function                                                                
  FAIL WebGL test #60: gl.getUniform(boolProgram, bval4Loc) should be true,false,true,false. Threw exception TypeError: gl.getUniform is not a function                                                          
  FAIL WebGL test #63: gl.getUniform(intProgram, ivalLoc) should be 1. Threw exception TypeError: gl.getUniform is not a function                                                                                
  FAIL WebGL test #64: gl.getUniform(intProgram, ival2Loc) should be 2,3. Threw exception TypeError: gl.getUniform is not a function                                                                             
  FAIL WebGL test #65: gl.getUniform(intProgram, ival3Loc) should be 4,5,6. Threw exception TypeError: gl.getUniform is not a function                                                                           
  FAIL WebGL test #66: gl.getUniform(intProgram, ival4Loc) should be 7,8,9,10. Threw exception TypeError: gl.getUniform is not a function                                                                        
  FAIL WebGL test #69: gl.getUniform(floatProgram, fvalLoc) should be 11. Threw exception TypeError: gl.getUniform is not a function                                                                             
  FAIL WebGL test #70: gl.getUniform(floatProgram, fval2Loc) should be 12,13. Threw exception TypeError: gl.getUniform is not a function                                                                         
  FAIL WebGL test #71: gl.getUniform(floatProgram, fval3Loc) should be 14,15,16. Threw exception TypeError: gl.getUniform is not a function                                                                      
  FAIL WebGL test #72: gl.getUniform(floatProgram, fval4Loc) should be 17,18,19,20. Threw exception TypeError: gl.getUniform is not a function                                                                   
  FAIL WebGL test #75: gl.getUniform(samplerProgram, s2DValLoc) should be 0. Threw exception TypeError: gl.getUniform is not a function                                                                          
  FAIL WebGL test #76: gl.getUniform(samplerProgram, sCubeValLoc) should be 1. Threw exception TypeError: gl.getUniform is not a function                                                                        
  FAIL WebGL test #79: gl.getUniform(matProgram, mval2Loc) should be 1,2,3,4. Threw exception TypeError: gl.getUniform is not a function                                                                         
  FAIL WebGL test #80: gl.getUniform(matProgram, mval3Loc) should be 5,6,7,8,9,10,11,12,13. Threw exception TypeError: gl.getUniform is not a function                                                           
  FAIL WebGL test #81: gl.getUniform(matProgram, mval4Loc) should be 14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29. Threw exception TypeError: gl.getUniform is not a function 
```
Are these tests that were legitimately expected to pass?

**Comment by gootorov** [comment]
No, `getUniform` is not implemented yet, so it's expected to fail. I guess the failures were raised because of the index thing. It should be fine now.

**Comment by gootorov** [comment]
Hmm, interesting, `getUniform` isn't set to fail in `gl-object-get-calls.html.ini`, but it now suddenly raises the errors, even though it should have always done that since it isn't implemented, I guess. I thought it raised the errors the previous time because of the index in `methods.ini`. These tests sometimes really confuse me :).

Anyway, updating the expectations in `gl-object-get-calls.html.ini` with that script should resolve the issue.

**Comment by brainlessdeveloper** [comment]
I placed the updates in a different commit because they have nothing to do with my PR

**Comment by brainlessdeveloper** [comment]
r?

File: tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/state/gl-object-get-calls.html.ini [hunk 6]
```diff
@@ -1,29 +1,53 @@
 [gl-object-get-calls.html]
-  expected: ERROR
-  [WebGL test #33: gl.getRenderbufferParameter(gl.RENDERBUFFER, gl.RENDERBUFFER_WIDTH) should be 2. Threw exception TypeError: gl.getRenderbufferParameter is not a function]
+  expected: TIMEOUT
```
Comment (gootorov) [inline tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/state/gl-object-get-calls.html.ini]: Looking at the last [log](http://build.servo.org/builders/linux-rel-wpt/builds/8557/steps/shell__4/logs/filtered-wpt-errorsummary.log), I guess this should be `OK`, then? Not sure why the script generated this to be a `TIMEOUT` and what's happening with this test in general.

cc @nox to the rescue.

Comment (brainlessdeveloper) [reply tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/state/gl-object-get-calls.html.ini]: IDK, I'm pushing again without the `TIMEOUT`

Comment (gootorov) [reply tests/wpt/mozilla/meta/webgl/conformance-1.0.3/conformance/state/gl-object-get-calls.html.ini]: Alright, let's see. Hope it works now :).

**Comment by brainlessdeveloper** [comment]
Finally :sweat_smile:

**Comment by gootorov** [comment]
> Test successful

cc @nox 
This can be r+'ed

File: components/script/dom/webglrenderingcontext.rs [hunk 7]
```diff
@@ -2348,6 +2348,45 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+    unsafe fn GetRenderbufferParameter(
+        &self,
+        _cx: *mut JSContext,
+        target: u32,
+        pname: u32
+    ) -> JSVal {
+        if self.bound_renderbuffer.get().is_none() {
+            self.webgl_error(InvalidOperation);
+            return NullValue();
+        }
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: This should probably move after the target and parameter names, to preserve the same order used in https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glGetRenderbufferParameteriv.xml.

**Review by emilio — APPROVED** [review]
Looks good modulo that. Let me know if you prefer to move that in a followup, that's fine for me :)

**Pull request merged**